### PR TITLE
Add x86-64 into the loaderarm branch

### PR DESCRIPTION
**Pull request by betahxy**
### Summary of this Pull Request (PR)

**Add description here.**

### Intent for your PR

Choose one (Mandatory):

- [x] This PR is for a code-review and is intended to get feedback, but not to be pulled yet.
- [ ] This PR is mature, and ready to be integrated into the repo.

### Reviewers (Mandatory):
(Specify @<github.com username(s)> of the reviewers. Ex: @user1, @user2)
@gparmer @WenyuanShao @msdx321 

### Code Quality

As part of this pull request, I've considered the following:

[Style](https://github.com/gparmer/composite/raw/ppos/doc/style_guide/composite_coding_style.pdf):

- [x] Comments adhere to the Style Guide (SG)
- [x] Spacing adhere's to the SG
- [x] Naming adhere's to the SG
- [x] All other aspects of the SG are adhered to, or exceptions are justified in this pull request
- [ ] I have run the auto formatter on my code before submitting this PR (see doc/auto_formatter.md for instructions)

[Code Craftsmanship](http://www2.seas.gwu.edu/~gparmer/posts/2016-03-07-code-craftsmanship.html):

- [x] I've made an attempt to remove all redundant code
- [x] I've considered ways in which my changes might impact existing code, and cleaned it up
- [x] I've formatted the code in an effort to make it easier to read (proper error handling, function use, etc...)
- [x] I've commented appropriately where code is tricky
- [x] I agree that there is no "throw-away" code, and that code in this PR is of high quality

### Testing

I've tested the code using the following test programs (provide list here):

- [ ] micro_booter
- [ ] unit_pingpong
- [ ] unit_schedtests
- [ ] ...(add others here)

### CODE REVIEW

**Comment by gparmer** [comment]
Please add Wenyuan, me and Bite to the listed reviewers.

**Comment by gparmer** [comment]
Rename this PR as something appropriate: "Add x86-64 into the loaderarm branch". There is no mention of the main core of the PR: x86-64!!!

**Comment by gparmer** [comment]
Have you tested that the 32 it x86 still works on this?

**Comment by betahxy** [comment]
> Have you tested that the 32 it x86 still works on this?

Not yet. You're right, I need to test it.

File: src/platform/x86/x86_64/gdt.c [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	      | (1 << 15)             	/* Present. */
+	      | (limit & 0xf0000)     	/* Limit 16:19. */
+	      | (base & 0xff000000)); 	/* Base 31:24. */
+
+	*tss_desc = (e0 | ((u64_t)e1 << 32));	
+	*(tss_desc + 1) = e2;
+}
+
+static void
+flash_selectors(void)
```
Comment (gparmer) [inline src/platform/x86/x86_64/gdt.c:158]: I don't know what this does from the name. The assembly isn't very self-documenting, so a good name, and potentially a comment is pretty necessary. What problem is this solving that isn't necessary on x86-32 (or is this adapted from there?).

File: src/platform/x86/x86_64/hpet.c [hunk 2]
```diff
@@ -0,0 +1,304 @@
+#include <thd.h>
```
Comment (gparmer) [inline src/platform/x86/x86_64/hpet.c:1]: Share this file with x86-32? symblink

File: src/platform/x86/x86_64/idt.c [hunk 3]
```diff
@@ -0,0 +1,174 @@
+#include "kernel.h"
```
Comment (gparmer) [inline src/platform/x86/x86_64/idt.c:1]: share with x86-32.

File: src/platform/x86/x86_64/isr.h [hunk 4]
```diff
@@ -0,0 +1,94 @@
+#ifndef ISR_H
```
Comment (gparmer) [inline src/platform/x86/x86_64/isr.h:1]: share with x86-32?

File: src/platform/x86/x86_64/kernel.c [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+	glb_memlayout.kern_end = &end;
+	assert((u64_t)&end % RETYPE_MEM_NPAGES * PAGE_SIZE == 0);
+
+	printk("Initial component found:\n");
+	/* These values have to be higher-half addresses */
+	glb_memlayout.mod_start = &_binary_constructor_start;
+	glb_memlayout.mod_end   = &_binary_constructor_end;
+
+	/* FIXME: set mod start and end temporarily in order to test */
+	//glb_memlayout.mod_start = &end + 1024;
```
Comment (gparmer) [inline src/platform/x86/x86_64/kernel.c:67]: remove commented out code.

If this FIXME is relevant, it should likely be fixed, I'd imagine.

File: src/platform/x86/x86_64/kernel.c [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+	unsigned int i = 0, wastage = 0;
+	struct multiboot_tag *tag;
+
+	glb_memlayout.allocs_avail = 1;
+
+	if (mboot_magic != MULTIBOOT2_BOOTLOADER_MAGIC) {
+		die("multiboot magic not correct\n");
+	}
+
+	if (mboot_addr & 7)
```
Comment (gparmer) [inline src/platform/x86/x86_64/kernel.c:53]: style (likely fixed with the format script).

File: src/platform/x86/x86_64/kernel.c [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+	}
+
+	size = 0;
+
+	printk("Memory regions:\n\n");
+	for (tag = (struct multiboot_tag *) (mboot_addr + 8);
+       tag->type != MULTIBOOT_TAG_TYPE_END;
+       tag = (struct multiboot_tag *) ((multiboot_uint8_t *) tag 
+                                       + ((tag->size + 7) & ~7)))
+    {
```
Comment (gparmer) [inline src/platform/x86/x86_64/kernel.c:85]: Indentation and style are off.

File: src/platform/x86/x86_64/kernel.c [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+		die("not found tag!\n");
+	}
+
+	size = 0;
+
+	printk("Memory regions:\n\n");
+	for (tag = (struct multiboot_tag *) (mboot_addr + 8);
+       tag->type != MULTIBOOT_TAG_TYPE_END;
+       tag = (struct multiboot_tag *) ((multiboot_uint8_t *) tag 
+                                       + ((tag->size + 7) & ~7)))
```
Comment (gparmer) [inline src/platform/x86/x86_64/kernel.c:84]: Consider isolating this stuff in an multiboot file. Is the 7 magic; perhaps a name would be helpful.

File: src/platform/x86/x86_64/kernel.h [hunk 9]
```diff
@@ -0,0 +1,103 @@
+#ifndef KERNEL_H
```
Comment (gparmer) [inline src/platform/x86/x86_64/kernel.h:1]: shared with 32?

File: src/platform/x86/x86_64/lapic.c [hunk 10]
```diff
@@ -0,0 +1,560 @@
+#include "kernel.h"
```
Comment (gparmer) [inline src/platform/x86/x86_64/lapic.c:1]: shared

File: src/platform/x86/x86_64/loader.S [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+.align 4
+.long MULTIBOOT_MAGIC
+.long MULTIBOOT_FLAGS
+.long MULTIBOOT_CHECKSUM
+
+.set STACKSIZE, 0x1000
+.comm stack, (STACKSIZE * NUM_CPU), 0x1000
+
+.globl loader
+loader:
```
Comment (gparmer) [inline src/platform/x86/x86_64/loader.S:21]: There are a lot of fixmes here. This seems like it needs to be cleaned up.

I'm a little surprized with the use of movq before switching to 64 bit, but I don't remember any of the details here, so it is likely OK.

File: src/platform/x86/x86_64/loader.S [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+
+.align 4096
+.code16
+.globl smppatchstart
+smppatchstart:
+.globl smploader
+smploader:
+	cli
+
+	movw    %cs, %ax
```
Comment (gparmer) [inline src/platform/x86/x86_64/loader.S:122]: When you get to fixing this, you have to remember that the APs boot up into 16 bit. So you can't just change instructions to `q` variants, and use 64 bit registers.

File: src/platform/x86/x86_64/mem_layout.h [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+ * mem_X_{end,start}, where X is:
+ * - kern  (K): the actual kernel image
+ * - bootc (B): the memory for the booter component
+ * - boot  (A): memory that is dynamically allocated at boot time (kernel bump allocator)
+ * - kmem:      all kernel memory at boot-up time represented in the retype table
+ * - utmem (U): memory that is untyped, and used by the components
+ *
+ * The memory image (physical memory):
+ *
+ * |-------|KKKKKKKK|WWWWW|BBBBBBBB|AA|UUUUUUUUUUUUUUUUUUU|
```
Comment (gparmer) [inline src/platform/x86/x86_64/mem_layout.h:25]: Validate that this image is up-to-date.

Perhaps share this file.

File: src/platform/x86/x86_64/miniacpi.c [hunk 14]
```diff
@@ -0,0 +1,354 @@
+#include "kernel.h"
```
Comment (gparmer) [inline src/platform/x86/x86_64/miniacpi.c:1]: share.

File: src/platform/x86/x86_64/multiboot.h [hunk 15]
```diff
@@ -0,0 +1,63 @@
+#ifndef MULTIBOOT_H
```
Comment (gparmer) [inline src/platform/x86/x86_64/multiboot.h:1]: share or remove?

Should we be using multiboot2 for x86-32 as well?

File: src/platform/x86/x86_64/multiboot2.h [hunk 16]
```diff
@@ -0,0 +1,380 @@
...
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL ANY
+ *  DEVELOPER OR DISTRIBUTOR BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ *  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ *  IN THE SOFTWARE.
+ */
+
+#ifndef _MULTIBOOT2_HEADER_H_
```
Comment (gparmer) [inline src/platform/x86/x86_64/multiboot2.h:23]: remove starting and trailing `_`. We don't do that consistently.

File: src/platform/x86/x86_64/ppos.sh [hunk 17]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
```
Comment (gparmer) [inline src/platform/x86/x86_64/ppos.sh:1]: remove.

File: src/platform/x86/x86_64/printk.c [hunk 18]
```diff
@@ -0,0 +1,51 @@
+#include "string.h"
```
Comment (gparmer) [inline src/platform/x86/x86_64/printk.c:1]: Shared? Can it be?

File: src/platform/x86/x86_64/qemu.sh [hunk 19]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
```
Comment (gparmer) [inline src/platform/x86/x86_64/qemu.sh:1]: remove, perhaps. I don't know if this is still necessary.

File: src/platform/x86/x86_64/serial.c [hunk 20]
```diff
@@ -0,0 +1,97 @@
+#define ENABLE_SERIAL
```
Comment (gparmer) [inline src/platform/x86/x86_64/serial.c:1]: shared?

File: src/platform/x86/x86_64/start.S [hunk 21]
```diff
@@ -0,0 +1,159 @@
+#include "chal/shared/cos_config.h"
```
Comment (gparmer) [inline src/platform/x86/x86_64/start.S:1]: If this replaces loader.S, then remove loader.s. However, there is some code in there that we likely want (smp).

I don't see a reason to not just move this into loader.S and keep the file names consistent with 32 bit.

File: src/platform/x86/x86_64/string.c [hunk 22]
```diff
@@ -0,0 +1,45 @@
+#include "string.h"
```
Comment (gparmer) [inline src/platform/x86/x86_64/string.c:1]: shared

File: src/platform/x86/x86_64/string.h [hunk 23]
```diff
@@ -0,0 +1,14 @@
+#ifndef STRING_H
```
Comment (gparmer) [inline src/platform/x86/x86_64/string.h:1]: shared.

File: src/platform/x86/x86_64/tss.c [hunk 24]
```diff
@@ -0,0 +1,19 @@
+#include "kernel.h"
```
Comment (gparmer) [inline src/platform/x86/x86_64/tss.c:1]: shared

File: src/platform/x86/x86_64/tss.h [hunk 25]
```diff
@@ -0,0 +1,24 @@
+#ifndef TSS_H
+#define TSS_H
+
+#include "shared/cos_types.h"
+struct tss
```
Comment (gparmer) [inline src/platform/x86/x86_64/tss.h:5]: style (`{` on the same line)

File: src/platform/x86/x86_64/user.c [hunk 26]
```diff
@@ -0,0 +1,2 @@
+#include "kernel.h"
```
Comment (gparmer) [inline src/platform/x86/x86_64/user.c:1]: unnecessar5y file, perhaps.

File: src/platform/x86/x86_64/vga.c [hunk 27]
```diff
@@ -0,0 +1,237 @@
+/*
```
Comment (gparmer) [inline src/platform/x86/x86_64/vga.c:1]: shared

File: src/platform/x86/x86_64/vtxprintf.c [hunk 28]
```diff
@@ -0,0 +1,304 @@
+/*  vtxprintf.c, from
```
Comment (gparmer) [inline src/platform/x86/x86_64/vtxprintf.c:1]: shared (after updating 32?)

File: src/platform/x86/x86_64/vtxprintf.h [hunk 29]
```diff
@@ -0,0 +1,39 @@
+/*
```
Comment (gparmer) [inline src/platform/x86/x86_64/vtxprintf.h:1]: Update 32 and share?

File: src/platform/x86/x86_64/vm.c [hunk 30]
```diff
@@ -0,0 +1,206 @@
...
+	u8_t *k;
+
+	assert((unsigned long)mem_bootc_start() % RETYPE_MEM_NPAGES == 0);
+	assert((unsigned long)mem_bootc_end() % RETYPE_MEM_NPAGES == 0);
+	for (k = mem_bootc_start(); k < mem_bootc_end(); k += PAGE_SIZE * RETYPE_MEM_NPAGES) {
+		if (retypetbl_retype2user((void *)chal_va2pa(k), PAGE_ORDER)) assert(0);
+	}
+}
+
+u8_t *mem_boot_alloc(int npages) /* boot-time, bump-ptr heap */
```
Comment (gparmer) [inline src/platform/x86/x86_64/vm.c:39]: `u8_t *` on separate line.

**Review by gparmer — CHANGES_REQUESTED** [review]
Most of the feedback is very small things. You did a really great job on this!

We do want to strongly consider sharing implementations with 32 bit x86 for a lot of this. Many files seem like they are identical, or lightly changed.

I've posted questions throughout the code review. You should consider the questions, but you should NOT answer them here. I won't be able to keep up with the conversation using github interfaces. If you want to discuss anything, we can chat on slack.

**Pull request closed without merging**